### PR TITLE
Add DhtPacket::get_payload

### PR DESCRIPTION
**Pull request by NamsooCho**
apply new DhtPacketPayload::from_bytes() to get_payload() of dht_impl.rs.
This PR has rebased on #11

### CODE REVIEW

File: src/toxcore/dht_new/dht_node.rs [hunk 1]
```diff
@@ -0,0 +1,1295 @@
...
+    }
+
+    /**
+    Remove nodes that have crossed `secs` timeout threshold.
+    */
+    // TODO: test
+    // TODO: add fn for ping/getn req timeouts with hardcoded consts?
+    pub fn remove_timed_out(&mut self, secs: u64) {
+        for pk in self.getn_timeout.get_timed_out(secs) {
+            debug!("Removing timed out node");
```
Comment (kurnevsky) [inline src/toxcore/dht_new/dht_node.rs]: I think pk should be added to the message, in other case it will be useless to have many same records in a row.

Comment (kpp) [reply src/toxcore/dht_new/dht_node.rs]: We should remove that code from Zetok. It is malformed.

File: src/toxcore/dht_new/dht_impl.rs [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+        let decrypted = open(&self.payload, &self.nonce, &self.pk,
+                            own_secret_key)
+            .and_then(|d| Ok(d))
+            .map_err(|e| {
+                debug!("Decrypting DhtPacket failed!");
+                Error::new(ErrorKind::Other,
+                    format!("DhtPacket decrypt error: {:?}", e))
+            });
+
+        match DhtPacketPayload::from_bytes(&decrypted.unwrap_or(vec![0]), self.packet_kind) {
```
Comment (kurnevsky) [inline src/toxcore/dht_new/dht_impl.rs]: Why unwrap_or(vec![0])? You already did .map_err(..) part so just add qustion mark to return error properly.

File: src/toxcore/dht_new/dht_impl.rs [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+                debug!("Decrypting DhtRequest failed!");
+                Error::new(ErrorKind::Other,
+                    format!("DhtRequest decrypt error: {:?}", e))
+            });
+
+        // size    | description
+        // 1       | nat ping req|resp (0xfe)
+        // 1       | Request or Response flag
+        // 8       | Request Id (Ping Id)
+        let packet_payload = decrypted.unwrap_or(vec![0,0]);
```
Comment (kurnevsky) [inline src/toxcore/dht_new/dht_impl.rs]: I'd propose to do here the same thing as for DhtPacket::get_payload to reduce code size.

File: src/toxcore/dht_new/dht_impl.rs [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+        }
+    }
+
+    /// Get precomputed keys
+    /// If the Key is not found in cache, create symmetric key and insert it into cache for later use.
+    pub fn get_symmetric_key (&mut self, key_pair: (&SecretKey, &PublicKey)) -> Result<PrecomputedKey, Error> {
+        let key = HashKeys(key_pair.0.clone(), key_pair.1.clone());
+        match self.cache.get(&key) {       // if symmetric key exists in cache, returns with the value
+            Some(k) => {
+                return Ok(k.clone());
```
Comment (kurnevsky) [inline src/toxcore/dht_new/dht_impl.rs]: Space at the end of line :)

**Comment by NamsooCho** [comment]
I reopen this PR with new contents.
New PR has only codec related codes.

File: src/toxcore/dht_new/packet.rs [hunk 5]
```diff
@@ -178,6 +179,49 @@ impl FromBytes for DhtPacket {
...
+        let decrypted = open(&self.payload, &self.nonce, &self.pk,
+                            own_secret_key)
+            .and_then(|d| Ok(d))
+            .map_err(|e| {
+                debug!("Decrypting DhtPacket failed!");
+                Error::new(ErrorKind::Other,
+                    format!("DhtPacket decrypt error: {:?}", e))
+            });
+
+        match DhtPacketPayload::from_bytes(&decrypted.unwrap_or(vec![0]), self.packet_kind) {
```
Comment (kurnevsky) [inline src/toxcore/dht_new/packet.rs]: Why unwrap_or(vec![0])? You already did .map_err(..) part so just add question mark to return error properly.

File: src/toxcore/dht_new/packet.rs [hunk 6]
```diff
@@ -178,6 +179,49 @@ impl FromBytes for DhtPacket {
...
+    - fails to decrypt
+    - fails to parse as given packet type
+    */
+    pub fn get_payload(&self, own_secret_key: &SecretKey) -> Result<Option<DhtPacketPayload>, Error>
+    {
+        debug!(target: "DhtPacket", "Getting packet data from DhtPacket.");
+        trace!(target: "DhtPacket", "With DhtPacket: {:?}", self);
+        let decrypted = open(&self.payload, &self.nonce, &self.pk,
+                            own_secret_key)
+            .and_then(|d| Ok(d))
```
Comment (kurnevsky) [inline src/toxcore/dht_new/packet.rs]: It seems this line is unnecessary.

File: src/toxcore/dht_new/packet.rs [hunk 7]
```diff
@@ -178,6 +179,49 @@ impl FromBytes for DhtPacket {
...
+
+    To get info about it's packet type use
+    [`.kind()`](./struct.DhtPacket.html#method.kind) method.
+
+    Returns `Error` in case of failure:
+
+    - fails to decrypt
+    - fails to parse as given packet type
+    */
+    pub fn get_payload(&self, own_secret_key: &SecretKey) -> Result<Option<DhtPacketPayload>, Error>
```
Comment (kurnevsky) [inline src/toxcore/dht_new/packet.rs]: Why Option here?

**Pull request merged**